iot/iot-device-registry-infinispan/src/main/java/io/enmasse/iot/registry/infinispan/util/DeviceRegistryTokenAuthProvider.java

```diff
@@ -36,17 +43,104 @@
 
     private final NamespacedKubernetesClient client;
     private final AuthApi authApi;
+    private final Vertx vertx;
+    private final EmbeddedCacheManager cacheManager = new DefaultCacheManager();
+    private Cache<String, TokenReview> tokens;
+    private Cache<String, Boolean> authorizations;
 
-    public DeviceRegistryTokenAuthProvider() {
+    public DeviceRegistryTokenAuthProvider(Vertx vertx, Duration tokenExpiration) {
+        log.info("Using token cache expiration of {}", tokenExpiration);
+        this.vertx = vertx;
         this.client = new DefaultKubernetesClient();
         this.authApi = new KubeAuthApi(this.client, this.client.getConfiguration().getOauthToken());
+        ConfigurationBuilder config = new ConfigurationBuilder();
+        config.expiration().lifespan(tokenExpiration.toSeconds(), TimeUnit.SECONDS);
+        cacheManager.defineConfiguration("tokens", config.build());
+        cacheManager.defineConfiguration("subjects", config.build());
+        tokens = cacheManager.getCache("tokens");
+        authorizations = cacheManager.getCache("subjects");
+
     }
 
     @Override
     public void authenticate(JsonObject authInfo, Handler<AsyncResult<User>> resultHandler) {
 
         final String token = authInfo.getString(TOKEN);
 
+        performTokenReview(token, tokenResult -> {
+            if (tokenResult.failed()) {
+                resultHandler.handle(Future.failedFuture(tokenResult.cause()));
+            }
+            TokenReview review = tokenResult.result();
+            if (review.isAuthenticated()) {
+                String role = getRole(authInfo, review, authApi);
```

Comment:
I think if getRole throws an exception there, then the result handler will never be called.